File: src/domain/repositories/file_repository.rs
use std::path::Path;

use anyhow::Result;
use async_trait::async_trait;

use crate::domain::models::file::File;

/// Repository interface for file operations
#[async_trait::async_trait]
pub trait FileRepository: Send + Sync {
    /// Get a file by its ID
    async fn get_file_by_id(&self, id: &str) -> Result<Option<File>>;
    
    /// Get a file by its path
    async fn get_file_by_path(&self, path: &str) -> Result<Option<File>>;
    
    /// Get all files in a folder
    async fn get_files_in_folder(&self, folder_id: &str) -> Result<Vec<File>>;
    
    /// Create a new file
    async fn create_file(&self, file: &File, content: &[u8]) -> Result<File>;
    
    /// Update file metadata
    async fn update_file(&self, file: &File) -> Result<File>;
    
    /// Delete a file
    async fn delete_file(&self, id: &str) -> Result<()>;
    
    /// Download file content
    async fn download_file(&self, id: &str) -> Result<Vec<u8>>;
    
    /// Upload file content
    async fn upload_file(&self, id: &str, content: &[u8]) -> Result<File>;
    
    /// Move a file to a different folder
    async fn move_file(&self, id: &str, new_parent_id: &str, new_name: Option<&str>) -> Result<File>;
    
    /// Copy a file to a different folder
    async fn copy_file(&self, id: &str, new_parent_id: &str, new_name: Option<&str>) -> Result<File>;
    
    /// Get favorite files
    async fn get_favorite_files(&self) -> Result<Vec<File>>;
    
    /// Set file as favorite or not
    async fn set_favorite(&self, id: &str, is_favorite: bool) -> Result<File>;
    
    /// Get recently modified files
    async fn get_recent_files(&self, limit: usize) -> Result<Vec<File>>;
    
    /// Create a file from a local path
    async fn import_file_from_path(&self, local_path: &Path, parent_id: &str) -> Result<File>;
    
    /// Export a file to a local path
    async fn export_file_to_path(&self, id: &str, local_path: &Path) -> Result<()>;
    
    /// Search for files by name
    async fn search_files(&self, query: &str) -> Result<Vec<File>>;
}

File: lib/domain/repositories/file_repository.dart
import 'dart:typed_data';
import '../entities/file_entity.dart';

abstract class FileRepository {
  /// List files in a folder (null = root).
  Future<List<FileEntity>> listFiles({String? folderId});

  /// Get file metadata by ID.
  Future<FileEntity> getFile(String id);

  /// Upload a file (simple upload for small files).
  Future<FileEntity> uploadFile({
    required String name,
    required String? folderId,
    required Stream<List<int>> fileStream,
    required int fileSize,
    required String mimeType,
  });

  /// Download a file. Returns bytes stream.
  Future<Stream<List<int>>> downloadFile(String id);

  /// Download file to a local path.
  Future<String> downloadFileToPath(String id, String localPath);

  /// Delete a file (soft-delete to trash).
  Future<void> deleteFile(String id);

  /// Rename a file.
  Future<FileEntity> renameFile(String id, String newName);

  /// Move a file to another folder.
  Future<FileEntity> moveFile(String id, String targetFolderId);

  /// Get thumbnail bytes.
  Future<Uint8List> getThumbnail(String id, {String size = '256'});
}
